zephyr/lib/alloc.c

```diff
@@ -154,6 +154,20 @@ static void *heap_alloc_aligned(struct k_heap *h, size_t min_align, size_t bytes
 	struct sys_memory_stats stats;
 #endif
 
+	/*
```

Comment:
problem:
cacheline is 64bytes long
zephyr puts metadata (8 bytes? no matter, lets assume 8) in front of allocated region

lets assume 
1) allocate 16 bytes (from cached area) by core 1 
from 0 to 7 - metadata (zepyr does use uncached alias __section(".heap_mem")  is in uncached area)
from 8 to 23 - data - translated to cached alias

2) core 1 is making changes in bytes 8-23,  at the moment in cache only

3) allocate 8 bytes (from cached area)  by core 2, as before
from 24 to 31 - metadata
from 32 - 40 - data

4) for any reason core 1 writes back data to RAM from the cache. **whole cacheline is written, heap metadata 23-41 is lost**

that exactly what was happening during my debug of the defect I mentioned

to avoid this there's a trick - always keep 1 separate cacheline between each allocation - just for metadata. It is done by aligning both start and length of every single allocation to cacheline size

**I not an author of this, I'm not claiming the algorithm is perfect. ** On contrary, I'm considering it as a dirty trick with enormous memory waste. Anyway, at the moment it is as it is. 
This PR is not trying to make it better, it is just making the trick work as it was design to work. And we need it quick because its affecting every multicore operation.

Reply:
Ack, thanks @marcinszkudlinski . So we need #7323 as urgent fix.

Reply:
ok, it is doing basically the same as my fix than, just at other point. 
Mine is safer (more tolerant to mistakes), yours - more memory friendly
Up to you what is to be closed 

Anyway - we DO need to make heap better, there's a lot of small allocations and using 2 cachelines for most of them is  something we simply cannot afford
FYI: @softwarecki @dabekjakub